src/policy/policy.cpp

```diff
@@ -154,6 +154,126 @@ bool AreInputsStandard(const CTransaction& tx, const CCoinsViewCache& mapInputs)
     return true;
 }
 
+BadTransaction IsBadWitness(const CTransaction& tx, const CCoinsViewCache& mapInputs)
```

Comment:
I don't really like that for two reasons:
1. Normally, `policy.cpp` is for mempool policy, but this one is part of consensus.
2. It seems to duplicate the logic of EvalScript which already detect correctly bad witnesses.

I think you should use `IsWitnessProgram` to detect if it is a witness, and if it is, try to call EvalScript and check if the error returns is the malleation problem. This may involve changing EvalScript function so that we don't execute anything after the malleated witness checks.